File: src/lib/submissionValue.ts
/**
 * Reads a value from a flat or nested submission payload by key.
 * Supports dot and slash notation for nested fields (e.g. "group/question").
 */
export function getPayloadValue(payload: Record<string, unknown>, key: string): unknown {
  const trimmed = key.trim();
  if (!trimmed) return undefined;

  if (Object.prototype.hasOwnProperty.call(payload, trimmed)) {
    return payload[trimmed];
  }

  const readNested = (segments: string[]): unknown => {
    let current: unknown = payload;
    for (const seg of segments) {
      if (!seg) continue;
      if (typeof current !== "object" || current === null || Array.isArray(current)) {
        return undefined;
      }
      if (!Object.prototype.hasOwnProperty.call(current, seg)) {
        return undefined;
      }
      current = (current as Record<string, unknown>)[seg];
    }
    return current;
  };

  if (trimmed.includes("/")) {
    const value = readNested(trimmed.split("/"));
    if (value !== undefined) return value;
  }
  if (trimmed.includes(".")) {
    const value = readNested(trimmed.split("."));
    if (value !== undefined) return value;
  }

  return undefined;
}

File: app/lib/submission_value.py
"""
Read a value from a flat or nested submission payload by key.
Supports dot and slash notation for nested fields (e.g. "group/question").
"""

from typing import Any


def get_payload_value(payload: dict, key: str) -> Any:
    trimmed = key.strip()
    if not trimmed:
        return None

    if trimmed in payload:
        return payload[trimmed]

    def read_nested(segments: list[str]) -> Any:
        current: Any = payload
        for seg in segments:
            if not seg:
                continue
            if not isinstance(current, dict):
                return None
            if seg not in current:
                return None
            current = current[seg]
        return current

    if '/' in trimmed:
        value = read_nested(trimmed.split('/'))
        if value is not None:
            return value

    if '.' in trimmed:
        value = read_nested(trimmed.split('.'))
        if value is not None:
            return value

    return None
